app/shared/text-editor/plugins/plugin-instance.tsx

```diff
@@ -0,0 +1,96 @@
+import { useEffect, useMemo } from "react";
+import {
+  useSubscribe,
+  useUserProps,
+  type Instance,
+  toCss,
+} from "@webstudio-is/sdk";
+import { primitives } from "../../canvas-components";
+import { $createInstanceNode, InstanceNode } from "../nodes/node-instance";
+import {
+  createCommand,
+  $isRangeSelection,
+  $getSelection,
+  COMMAND_PRIORITY_EDITOR,
+  useLexicalComposerContext,
+  type LexicalCommand,
+} from "../lexical";
+import { useCss } from "~/canvas/features/wrapper-component/use-css";
+import { useBreakpoints } from "~/shared/nano-states";
+
+const INSERT_INSTANCE_COMMAND: LexicalCommand<CreateInstancePayload> =
+  createCommand();
+
+type CreateInstancePayload = {
+  instance: Instance;
+  props: Record<string, unknown>;
+};
+
+export const InstancePlugin = () => {
+  const [editor] = useLexicalComposerContext();
+  useEffect(() => {
+    if (editor.hasNodes([InstanceNode]) === false) {
+      throw new Error("InstancePlugin: InstanceNode not registered on editor");
+    }
+
+    return editor.registerCommand<CreateInstancePayload>(
+      INSERT_INSTANCE_COMMAND,
+      ({ instance, props }) => {
+        const selection = $getSelection();
+        const text = selection?.getTextContent();
+        if ($isRangeSelection(selection) && text) {
+          const instanceNode = $createInstanceNode({
+            component: (
+              <InlineWrapperComponent {...props} instance={instance}>
+                {text}
+              </InlineWrapperComponent>
+            ),
+            text,
+          });
+          selection.insertNodes([instanceNode]);
+        }
+
+        return true;
+      },
+      COMMAND_PRIORITY_EDITOR
+    );
+  }, [editor]);
+
+  useSubscribe<"insertInlineInstance", CreateInstancePayload>(
+    "insertInlineInstance",
+    (payload) => {
+      editor.dispatchCommand<CreateInstancePayload>(
+        INSERT_INSTANCE_COMMAND,
+        payload
+      );
+    }
+  );
+  return null;
+};
+
+const InlineWrapperComponent = ({
+  instance,
+  ...rest
+}: {
+  instance: Instance;
+  children: string;
+}) => {
+  const [breakpoints] = useBreakpoints();
+  const css = useMemo(
+    () => toCss(instance.cssRules, breakpoints),
+    [instance, breakpoints]
+  );
+  const className = useCss({ instance, css });
```

Comment:
This feel counter-intuitive. the rest of SDK handles this internally, but this component does it here. Is this intentional?

Reply:
wrapper component is part of the designer, not SDK, I had to create a new one for inline components inside the instance, because you don't want there the same wrapper component that renders it from outside, basically this plugin is all about integrating lexical with designer/canvas, not sure where is a better place for this, because its only ever going to be used inside the text editor.

I will see if we can reuse majority of this logic  as a single hook  so that there is less duplication